### DISCUSSION TITLE
List handling with SIMATIC AX in ST

### DISCUSSION THREAD

**Discussion by sjuergen** [simatic-ax/.discussions#18] [Ideas]
Hey guys, 

I implemented some [functions for list handling](https://github.com/simatic-ax/collections) (LinkedList, Fifo, Stack) with SIMATIC AX. But there are some limitations:
- ST does not support dynamic memory allocation
- ST does not support 'generics'
- There's no datatype like "Variant" available

So this [library](https://github.com/simatic-ax/collections) is a try to offer such functionality with that above mentioned limitations.

Actually im not so happy with that solution. Because If you enqueue something into the FIFO your object will not be copied. that means you've to manage all your instances of your array (example: enqueue an item and dequeue it, then you've to care, that this item can be "reused").
Another problem ist, that the following example does not work with ST:
```
CLASS Fifo IMPLEMENTS ICollection
  VAR PUBLIC
    buffer : REF_TO ARRAY[*] OF IListItem
  END_VAR
...
END_CLASS

CLASS myConreteItem IMPLEMENTS IListItem
...
END_CLASS

VAR_GLOBAL
  myfifo : Fifo  := (buffer := REF(items); // can't assign ARRAY of myConreteItem to ARRAY of IListItems.
  items : ARRAY[1..100] OF myConreteItem;  
END_VAR
```

I'm open for a discussion and ideas, how to handle "generic" list handling with SIMATIC AX.

**Comment by PTKu** [comment]
This looks nice will give it a try. I've heard there was a proposal for templates/generics in IEC group, but some members put it down; it may be time to reopen that discussion. 
Will play around with the library... I may get some ideas 😊

**Reply by yyliveyy** [reply]
As far as I know, the new standard IEC 61131-3 will be released next year. If the new standard does not have a description related to 'generics', it would be a very frustrating thing. Suggest adding this language feature to Simatic Ax.
